### ISSUE TITLE
kafka exactly once sink: transaction timeout maybe hoses sink

### ISSUE THREAD

**Issue by cjubb39** [MaterializeInc/materialize#10876]
### What version of Materialize are you using?

master

### How did you install Materialize?

Built from source

### What was the issue?

https://github.com/MaterializeInc/product/issues/136

- Kafka transaction timeout causes one of our transaction state transitions to trip an assert in rdkafka: https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_txnmgr.c#L237.  We've seen this occur in a customer running v0.12 (which predates some pretty big sink refactors)

I'm not sure that we handle this well so the purpose of this ticket is to:
1) Investigate and determine if this is still an issue
2) Fix it if necessary

**Comment by cjubb39** [comment]
This has occurred again with a customer running v0.20.0:

```
2022-03-08T03:42:49.724124Z ERROR librdkafka: TXNERR [thrd:main]: Current transaction failed in state CommittingTransaction: Transactional API operation (commit_transaction) timed out (_TIMED_OUT)
2022-03-08T03:42:49.725947Z  INFO dataflow::sink::kafka: Error requiring abort in kafka sink: RDKafkaError(Transactional API operation (commit_transaction) timed out)
```

and then sometime afterwards, there's a fencing error.

```
2022-03-08T04:38:51.888454Z ERROR kafka_util::client: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 10/10 brokers are down
2022-03-08T04:38:53.046492Z ERROR librdkafka: TXNERR [thrd:main]: Fatal transaction error: Failed to end transaction: Local: This instance has been fenced by a newer instance (_FENCED)
2022-03-08T04:38:53.046674Z ERROR librdkafka: FATAL [thrd:main]: Fatal error: Local: This instance has been fenced by a newer instance: Failed to end transaction: Local: This instance has been fenced by a newer instance
2022-03-08T04:38:53.047277Z ERROR kafka_util::client: librdkafka: Global error: Fatal (Local: Fatal error): Fatal error: Local: This instance has been fenced by a newer instance: Failed to end transaction : Local: This instance has been fenced by a newer instance
2022-03-08T04:39:06.003407Z  INFO dataflow::sink::kafka: shutting down kafka sink: kafka-u5
```

I believe I understand the transaction timeout error but I'm not immediately sure how we end up getting a fencing error that doesn't show up right after startup.  Because the initial error occurs while waiting for a `CommitTxn`, we know we've definitely started up and initialized by the time we got here (This version has most of the recent sink rewrites -- including the change to the fencing ordering)

**Comment by cjubb39** [comment]
From a user:
>  Hey, unsure if related but I have managed to replicate the error locally by accident.
It happened when the number the replication was less than the min.insync. I have seen similar errors like this with debezium when the acks=all was unable to be satisfied since the min.insync of 2 would not be achieved with 1 replication.

**Comment by philip-stoev** [comment]
I am getting the fencing error if there are two sinks writing to the same topic.

**Comment by cjubb39** [comment]
oh nice -- thanks for the repro philip

**Comment by bkirwi** [comment]
The shared output topic error has been addressed in #14364. But that doesn't explain the fencing error in this issue, I think. I suppose it's possible that another process started up or a thread revived itself and started trying to send data after an hour? Strange either way.

**Comment by cjubb39** [comment]
I think this should be addressed by #14965 which will crash and restart the sink when we hit this type of error.  It may not fix the underlying issue per se but, since we believe this type of error is transient, the sink will no longer be hosed if we restart
